src/any_agent/callbacks/span_generation/base.py

```diff
@@ -93,59 +90,43 @@ def _set_tool_input(
         call_id: str | None = None,
     ) -> Context:
         tracer = context.tracer
+        span = tracer.start_span(f"execute_tool {name}")
+
+        attributes = {
+            GenAI.OPERATION_NAME: "execute_tool",
+            GenAI.TOOL_NAME: name,
+        }
 
-        span = tracer.start_span(
-            name=f"execute_tool {name}",
-        )
-        span.set_attributes(
-            {
-                GenAI.OPERATION_NAME: "execute_tool",
-                GenAI.TOOL_NAME: name,
-            }
-        )
         if description is not None:
-            span.set_attribute(GenAI.TOOL_DESCRIPTION, description)
+            attributes[GenAI.TOOL_DESCRIPTION] = description
         if args is not None:
-            span.set_attribute(
-                GenAI.TOOL_ARGS,
-                json.dumps(
-                    args,
-                    default=str,
-                    ensure_ascii=False,
-                ),
-            )
+            attributes[GenAI.TOOL_ARGS] = self._serialize_for_attribute(args)
         if call_id is not None:
-            span.set_attribute("gen_ai.tool.call.id", call_id)
+            attributes["gen_ai.tool.call.id"] = call_id
 
+        span.set_attributes(attributes)
         context.current_span = span
-
         return context
 
+    def _determine_tool_status(self, tool_output: str) -> Status | StatusCode:
+        """Determine the status based on tool output content."""
+        if "Error calling tool:" in tool_output:
+            return Status(StatusCode.ERROR, description=tool_output)
+        return StatusCode.OK
+
     def _set_tool_output(self, context: Context, tool_output: Any) -> Context:
         span = context.current_span
 
         if tool_output is None:
             tool_output = "{}"
 
-        status: Status | StatusCode = StatusCode.OK
-
-        if isinstance(tool_output, str):
-            try:
-                json.loads(tool_output)
-                output_type = "json"
-            except json.decoder.JSONDecodeError:
-                output_type = "text"
-                if "Error calling tool:" in tool_output:
-                    status = Status(StatusCode.ERROR, description=tool_output)
-        else:
-            try:
-                tool_output = json.dumps(tool_output, default=str, ensure_ascii=False)
-                output_type = "json"
-            except TypeError:
-                tool_output = str(tool_output)
-                output_type = "text"
+        output_attr, output_type = self._determine_output_type(tool_output)
+        status = (
+            self._determine_tool_status(output_attr)
+            if output_type == "text"
+            else StatusCode.OK
```

Comment:
💅 since we are refactoring, I would put this `output_type` check inside `_determine_tool_status`